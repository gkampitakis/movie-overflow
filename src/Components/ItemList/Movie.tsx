import React from 'react';
import { Movie } from '../../types';
import { getImage } from '../../utils';

export function MovieItem(movie: Movie) {
  return (
    <>
      <img
        src={getImage(movie.poster_path, '185')}
        loading="lazy"
        alt={movie.original_title}
      />
      <div className="details">
        <div>
          <div className="header">
            <h2>{movie.original_title}</h2>
            <p>Movie</p>
          </div>
          <p>{movie.release_date}</p>
        </div>
        <p className="overview">{movie.overview}</p>
      </div>
    </>
  );
}
