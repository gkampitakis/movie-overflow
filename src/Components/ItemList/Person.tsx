import React from 'react';
import { Person } from '../../types';
import { getImage, avatar } from '../../utils';

export function PersonItem(person: Person) {
  return (
    <>
      <img
        src={getImage(person.profile_path, '185', avatar(person.name, '185'))}
        loading="lazy"
        alt={person.name}
      />
      <div className="details">
        <div>
          <div className="header">
            <h2>{person.name}</h2>
            <p>Person</p>
          </div>
          <p>{person.known_for_department}</p>
        </div>
        <p className="overview">
          {person.known_for.map((t) => t.original_title || t.name).join(', ')}
        </p>
      </div>
    </>
  );
}
