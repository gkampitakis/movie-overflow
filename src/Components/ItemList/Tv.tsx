import React from 'react';
import { Series } from '../../types';
import Placeholder from '../../assets/placeholder.jpeg';

const imgSrc = (path: string) =>
  path ? `https://image.tmdb.org/t/p/w185/${path}` : Placeholder;

export function TvItem(tv: Series) {
  return (
    <>
      <img src={imgSrc(tv.poster_path)} loading="lazy" alt={tv.name} />
      <div className="details">
        <div>
          <div className="header">
            <h2>{tv.original_name}</h2>
            <p>Tv Series</p>
          </div>
          <p>{tv.first_air_date}</p>
        </div>
        <p className="overview">{tv.overview}</p>
      </div>
    </>
  );
}
