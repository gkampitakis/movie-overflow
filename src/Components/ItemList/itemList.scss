.item_list {
  cursor: pointer;
  display: flex;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
  border: 1px solid rgba(227, 227, 227, 1);
  border-radius: 8px;
  width: 95%;
  height: 180px;
  margin: 20px auto;
  transition: transform 0.5s;

  img {
    height: 100%;
    border-top-left-radius: 10px;
    border-bottom-left-radius: 10px;
    max-width: 120px;
  }

  .details {
    margin: 10px 20px;
    display: flex;
    flex-direction: column;
    justify-content: space-between;

    h2 {
      font-size: 1.5rem;
      letter-spacing: 2px;
      font-weight: 600;
    }

    p {
      color: #999;
    }

    .overview {
      color: black;
      font-size: 17px;
    }
  }

  &:hover {
    -webkit-transform: scale(1.02);
    transform: scale(1.02);
  }
}
