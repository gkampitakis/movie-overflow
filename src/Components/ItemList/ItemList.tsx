import React from 'react';
import { Actor, Movie, Series } from '../../types';
import NotFound from '../../assets/404.svg';
import './itemList.scss';

interface ItemListProps {
  item: Actor | Movie | Series;
}

export default function ItemList(props: ItemListProps) {
  const items = {
    movie: MovieItem,
    tv: TvItem,
    person: PersonItem
  };

  return items[props.item.media_type](props.item as any);
}

function PersonItem(actor: Actor) {
  const src = actor.profile_path
    ? `https://image.tmdb.org/t/p/w185/${actor.profile_path}`
    : NotFound;

  return (
    <article className="item_list">
      <img src={src} alt={actor.name} />
      <div className="details">
        <div>
          <h2>{actor.name}</h2>
          <p>{actor.known_for_department}</p>
        </div>
        <p className="overview">
          {actor.known_for.map((t) => t.original_title).join('*')}
        </p>
      </div>
    </article>
  );
}

function TvItem(tv: Series) {
  const src = tv.poster_path
    ? `https://image.tmdb.org/t/p/w185/${tv.poster_path}`
    : NotFound;

  return (
    <article className="item_list">
      <img src={src} alt={tv.name} />
      <div className="details">
        <div>
          <h2>{tv.original_name}</h2>
          <p>{tv.first_air_date}</p>
        </div>
        <p className="overview">{tv.overview}</p>
      </div>
    </article>
  );
}

function MovieItem(movie: Movie) {
  const src = movie.poster_path
    ? `https://image.tmdb.org/t/p/w185/${movie.poster_path}`
    : NotFound;

  return (
    <article className="item_list">
      <img src={src} alt={movie.original_title} />
      <div className="details">
        <div>
          <h2>{movie.original_title}</h2>
          <p>{movie.release_date}</p>
        </div>
        <p className="overview">{movie.overview}</p>
      </div>
    </article>
  );
}
