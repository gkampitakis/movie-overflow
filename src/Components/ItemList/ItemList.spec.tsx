import React from 'react';
import ItemList from '.';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {
  mockHistoryPush,
  mockPerson,
  mockMovie,
  mockTvSeries
} from '../../setupTests';

describe('ItemList', () => {
  describe('Person item', () => {
    it('Should render consistently', () => {
      const { container } = render(<ItemList item={mockPerson} />);

      expect(container).toMatchSnapshot();
    });

    it('Should render placeholder image', () => {
      mockPerson.profile_path = '';
      const { container } = render(<ItemList item={mockPerson} />);

      expect(container).toMatchSnapshot();
    });

    it('Should navigate to person details', () => {
      render(<ItemList item={mockPerson} />);

      userEvent.click(screen.getByTestId('item-list'));

      expect(mockHistoryPush).toHaveBeenCalledWith({
        pathname: '/person/1'
      });
    });
  });

  describe('Movie item', () => {
    it('Should render consistently', () => {
      const { container } = render(<ItemList item={mockMovie} />);

      expect(container).toMatchSnapshot();
    });

    it('Should render placeholder image', () => {
      mockMovie.poster_path = '';
      const { container } = render(<ItemList item={mockMovie} />);

      expect(container).toMatchSnapshot();
    });

    it('Should navigate to person details', () => {
      render(<ItemList item={mockMovie} />);

      userEvent.click(screen.getByTestId('item-list'));

      expect(mockHistoryPush).toHaveBeenCalledWith({
        pathname: '/movie/2'
      });
    });
  });

  describe('Tv item', () => {
    it('Should render consistently', () => {
      const { container } = render(<ItemList item={mockTvSeries} />);

      expect(container).toMatchSnapshot();
    });

    it('Should render placeholder image', () => {
      mockTvSeries.poster_path = '';
      const { container } = render(<ItemList item={mockTvSeries} />);

      expect(container).toMatchSnapshot();
    });

    it('Should navigate to person details', () => {
      render(<ItemList item={mockTvSeries} />);

      userEvent.click(screen.getByTestId('item-list'));

      expect(mockHistoryPush).toHaveBeenCalledWith({
        pathname: '/tv/3'
      });
    });
  });
});
