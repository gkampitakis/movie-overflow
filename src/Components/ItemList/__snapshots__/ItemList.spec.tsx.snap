// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemList Movie item Should render consistently 1`] = `
<div>
  <article
    class="item_list"
    data-testid="item-list"
  >
    <img
      alt="mockTitle"
      loading="lazy"
      src="https://image.tmdb.org/t/p/w185/mock_path"
    />
    <div
      class="details"
    >
      <div>
        <div
          class="header"
        >
          <h2>
            mockTitle
          </h2>
          <p>
            Movie
          </p>
        </div>
        <p>
          mockDate
        </p>
      </div>
      <p
        class="overview"
      >
        mockOverview
      </p>
    </div>
  </article>
</div>
`;

exports[`ItemList Movie item Should render placeholder image 1`] = `
<div>
  <article
    class="item_list"
    data-testid="item-list"
  >
    <img
      alt="mockTitle"
      loading="lazy"
      src="placeholder.jpeg"
    />
    <div
      class="details"
    >
      <div>
        <div
          class="header"
        >
          <h2>
            mockTitle
          </h2>
          <p>
            Movie
          </p>
        </div>
        <p>
          mockDate
        </p>
      </div>
      <p
        class="overview"
      >
        mockOverview
      </p>
    </div>
  </article>
</div>
`;

exports[`ItemList Person item Should render consistently 1`] = `
<div>
  <article
    class="item_list"
    data-testid="item-list"
  >
    <img
      alt="mockName"
      loading="lazy"
      src="https://image.tmdb.org/t/p/w185/mock_path"
    />
    <div
      class="details"
    >
      <div>
        <div
          class="header"
        >
          <h2>
            mockName
          </h2>
          <p>
            Person
          </p>
        </div>
        <p>
          Actor
        </p>
      </div>
      <p
        class="overview"
      >
        mockTvSeries, mockMovie
      </p>
    </div>
  </article>
</div>
`;

exports[`ItemList Person item Should render placeholder image 1`] = `
<div>
  <article
    class="item_list"
    data-testid="item-list"
  >
    <img
      alt="mockName"
      loading="lazy"
      src="https://eu.ui-avatars.com/api/?size=185&name=mockName"
    />
    <div
      class="details"
    >
      <div>
        <div
          class="header"
        >
          <h2>
            mockName
          </h2>
          <p>
            Person
          </p>
        </div>
        <p>
          Actor
        </p>
      </div>
      <p
        class="overview"
      >
        mockTvSeries, mockMovie
      </p>
    </div>
  </article>
</div>
`;

exports[`ItemList Tv item Should render consistently 1`] = `
<div>
  <article
    class="item_list"
    data-testid="item-list"
  >
    <img
      alt="mockTitle"
      loading="lazy"
      src="https://image.tmdb.org/t/p/w185/mock_path"
    />
    <div
      class="details"
    >
      <div>
        <div
          class="header"
        >
          <h2>
            mockTitle
          </h2>
          <p>
            Tv Series
          </p>
        </div>
        <p />
      </div>
      <p
        class="overview"
      >
        mockOverview
      </p>
    </div>
  </article>
</div>
`;

exports[`ItemList Tv item Should render placeholder image 1`] = `
<div>
  <article
    class="item_list"
    data-testid="item-list"
  >
    <img
      alt="mockTitle"
      loading="lazy"
      src="placeholder.jpeg"
    />
    <div
      class="details"
    >
      <div>
        <div
          class="header"
        >
          <h2>
            mockTitle
          </h2>
          <p>
            Tv Series
          </p>
        </div>
        <p />
      </div>
      <p
        class="overview"
      >
        mockOverview
      </p>
    </div>
  </article>
</div>
`;
